attestation-evidence-broker/docs/design/design.md

```diff
@@ -0,0 +1,122 @@
+
+| Contents of SEV(-ES) Attestation Evidence | SEV(-ES) firmware command |
+| :-: | :-: |
+| attestation report                        | ATTESTATION               |
+| CEK                                       | GET_ID                    |
+| OCA,PEK                                   | PDH_CERT_EXPORT           |
+
+## The methods of query SEV(-ES) Attestation Evidence
+
+According to the [officially feedback](https://github.com/AMDESE/AMDSEV/issues/71#issuecomment-926118314), **SEV(-ES) firmware APIs don't support query attestation report in SEV(-ES) guest and there is no plan to support it in the future. Instead, this capability will be available in SEV-SNP.**
+
+In some scenarios, the guest application needs to query the attestation report to establish a trusted channel with the remote peer. There are two approaches for a guest application to query an attestation evidence:
+
+- Hypercall approach
+- Vsock approach
+
+Considering time and cost, we only need to implement one of them.
+
+### Hypercall approach
+
+SEV(-ES) guest exits to VMM using `hypercall` and then interacts with PSP firmware to query attestation evidence, including attestation report, PEK, CEK, OCA certificate. In general, it includes the following steps:
+
+1. The guest application requests a shared memory page, initiates a hypercall, and switches from VMX non-root mode to VMX root mode.
+2. In VMX root mode, KVM sends the `GET_ID, PDH_CERT_EXPORT, ATTESTATION` command requests to PSP firmware.
+3. The shared memory page is filled with the data returned by the PSP firmware.
+4. The guest application can obtain attestation evidence by reading the data in the shared memory.
+
+Although this method can meet our requirements, it requires a lot of modifications to the guest kernel and KVM, and it is hard to be accepted by upstream.
+
+### Vsock approach
+
+In the current implementation, QEMU provides the QMP interface "query-sev-attestation-report" to query the attestation report in the host. The report is signed by PEK and contains the SHA-256 digest of the guest memory and VMSA encrypted with the LAUNCH_UPDATE. However, QEMU is not the only VMM. In order to support various VMM in different scenarios, it is necessary to design a general design to query attestation evidence from the host.
+
+The workflow of AEB is as followed:
+
+![architecture.png](architecture.png)
+
+1. The user-level application in the guest sends a request (including guest firmware handle) to AEB through Vsock.
+2. AEB requests to query attestation report, PEK, CEK, OCA certificate by calling multiple PSP firmware APIs (refer to the table above for specific API commands) and assembles these information into attestation evidence.
+3. AEB returns attestation evidence to the application in the guest.
+
+To query the attestation report in host with AEB, you need to apply the following patches, which will be sent to the community for review afterward:
```

Comment:
s/To query the attestation report in host with AEB, you need to apply the following patches, which will be sent to the community for review afterward:/In order to work out AEB, please manually apply [these patches](url to hack directory in github repo). Here are the workflows:/